inlanecruising_plugin/src/calculation.cpp

```diff
@@ -0,0 +1,308 @@
+
+namespace inlanecruising_plugin
+{
+
+std::vector<PointSpeedPair> points_in_time_boundary(const std::vector<PointSpeedPair>& points, double time_span) {
+  std::vector<PointSpeedPair> output;
+  output.reserve(points.size());
+  double current_time = 0;
+  PointSpeedPair prev_pair;
+  bool first = true;
+  for (const auto& p : points) {
+    if (current_time > time_span) {
+      return output;
+    }
+
+   ROS_WARN_STREAM("Evaluating time of point: " << std::get<0>(p).x() << ", " << std::get<0>(p).y());
+
+
+    if (first) {
+      output.push_back(p);
+      prev_pair = p;
+      first = false;
+      continue;
+    }
+
+    double delta_d = lanelet::geometry::distance2d(std::get<0>(p), std::get<0>(prev_pair));
+
+    current_time += delta_d / std::get<1>(prev_pair);
+
+    output.push_back(p);
+    prev_pair = p;
+  }
+
+  return output;
+}
+
+std::vector<PointSpeedPair> maneuvers_to_points(const std::vector<cav_msgs::Maneuver>& maneuvers,
+                                                double max_starting_downtrack, const carma_wm::WorldModelConstPtr& wm)
+{
+  std::vector<PointSpeedPair> points_and_target_speeds;
+  std::unordered_set<lanelet::Id> visited_lanelets;
+
+  bool first = true;
+  ROS_WARN_STREAM("VehDowntrack: " << max_starting_downtrack);
+  for (const auto& manuever : maneuvers)
+  {
+    if (manuever.type != cav_msgs::Maneuver::LANE_FOLLOWING)
+    {
+      throw std::invalid_argument("In-Lane Cruising does not support this maneuver type");
+    }
+
+    cav_msgs::LaneFollowingManeuver lane_following_maneuver = manuever.lane_following_maneuver;
+
+    double starting_downtrack = lane_following_maneuver.start_dist;
+    if (first) {
+      if (starting_downtrack > max_starting_downtrack) {
+        starting_downtrack = max_starting_downtrack;
+      }
+      first = false;
+    }
+
+    ROS_WARN_STREAM("Used downtrack: " << starting_downtrack);
+
+    auto lanelets = wm->getLaneletsBetween(starting_downtrack, lane_following_maneuver.end_dist, true);
+
+    ROS_WARN_STREAM("Maneuver");
+    std::vector<lanelet::ConstLanelet> lanelets_to_add;
+    for (auto l: lanelets) {
+      ROS_WARN_STREAM("Lanelet ID: " << l.id());
+      if (visited_lanelets.find(l.id()) == visited_lanelets.end()) {
+        lanelets_to_add.push_back(l);
+        visited_lanelets.insert(l.id());
+      }
+    }
+    
+    lanelet::BasicLineString2d route_geometry = carma_wm::geometry::concatenate_lanelets(lanelets_to_add);
+
+    bool first = true;
+    for (auto p : route_geometry)
+    {
+      if (first && points_and_target_speeds.size() != 0) {
+        first = false;
+        continue; // Skip the first point if we have already added points from a previous maneuver to avoid duplicates
+      }
+      points_and_target_speeds.push_back(std::make_pair(p, lane_following_maneuver.end_speed));
+    }
+  }
+
+  return points_and_target_speeds;
+}
+
+std::vector<PointSpeedPair> downsample_points(const std::vector<PointSpeedPair>& points, int nth_point)
```

Comment:
At some point we really ought to refactor some of this functionality using templates, so that we have one generic "downsample" function in a utils package somewhere.

Reply:
I can probably whip that up during my clean up work.